include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1386 @@
+
+	float cosFactor = is_bsdf ? abs(s.NdotL):max(s.NdotL,0.0);
+
+	irr_glsl_AnisotropicMicrofacetCache uf;
+	//here actually using stronger check for BSDF because it's probably worth it
+	if (op_isBSDF(op) && irr_glsl_isTransmissionPath(currInteraction.isotropic.NdotV, s.NdotL))
+	{
+		float orientedEta, rcpOrientedEta;
+		irr_glsl_getOrientedEtas(orientedEta, rcpOrientedEta, currInteraction.isotropic.NdotV, ior[0].x);
+		const bool valid = irr_glsl_calcAnisotropicMicrofacetCache(uf, true, currInteraction.isotropic.V.dir, s.L, currInteraction.T, currInteraction.B, currInteraction.isotropic.N, s.NdotL, s.VdotL, orientedEta, rcpOrientedEta);
+		//assert(valid);
+	}
+	else
+	{
+		uf = _uf;
+	}
+
+	if (cosFactor>FLT_MIN && op_hasSpecular(op))//does cos>0 check even has any point here? L comes from a sample..
+	{
+#ifdef OP_DIFFUSE
+		if (op==OP_DIFFUSE) {
+			pdf = irr_glsl_oren_nayar_pdf_wo_clamps(cosFactor);
+		} else
+#endif
+
+#ifdef OP_DIFFTRANS
+		if (op==OP_DIFFTRANS) {
+			pdf = irr_glsl_lambertian_transmitter_pdf_wo_clamps(cosFactor);
+		}
+		else
+#endif //OP_DIFFTRANS
+
+#ifdef NDF_GGX
+		if (ndf==NDF_GGX) {
+
+#if defined(OP_THINDIELECTRIC) || defined(OP_DIELECTRIC)
+			if (is_bsdf) {
+#ifdef ALL_ISOTROPIC_BXDFS
+				bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_dielectric_cos_eval_and_pdf(pdf, s, currInteraction.isotropic, uf.isotropic, ior[0].x, a2);
+#else
+				bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_dielectric_cos_eval_and_pdf(pdf, s, currInteraction.isotropic, uf.isotropic, ior[0].x, a2);//TODO aniso
+#endif
+			}
+			else
+#endif
+			{
+#ifdef ALL_ISOTROPIC_BXDFS
+				bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_cos_eval_DG(s, uf.isotropic, currInteraction.isotropic, a2);
+				pdf = irr_glsl_ggx_pdf(currInteraction.isotropic, uf.isotropic, a2);
+#else
+				bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_aniso_cos_eval_DG(s, uf, currInteraction, a, a2, ay, ay2);
+				pdf = irr_glsl_ggx_pdf(currInteraction, uf, a, ay, a2, ay2);
+#endif
+			}
+		} else
+#endif
+
+#ifdef NDF_BECKMANN
+		if (ndf==NDF_BECKMANN) {
+
+#if defined(OP_THINDIELECTRIC) || defined(OP_DIELECTRIC)
+			if (is_bsdf) {
+#ifdef ALL_ISOTROPIC_BXDFS
+				bxdf_eval_scalar_part = irr_glsl_beckmann_smith_height_correlated_dielectric_cos_eval_and_pdf(pdf, s, currInteraction.isotropic, uf.isotropic, ior[0].x, a2);
+#else
+				bxdf_eval_scalar_part = irr_glsl_beckmann_aniso_smith_height_correlated_dielectric_cos_eval_and_pdf(pdf, s, currInteraction, uf, ior[0].x, a, ay);
+#endif
+			}
+			else
+#endif
+			{
+#ifdef ALL_ISOTROPIC_BXDFS
+				bxdf_eval_scalar_part = irr_glsl_beckmann_height_correlated_cos_eval_DG(s, uf.isotropic, currInteraction.isotropic, a2);
+				pdf = irr_glsl_beckmann_pdf(currInteraction.isotropic, uf.isotropic, a2);
+#else
+				bxdf_eval_scalar_part = irr_glsl_beckmann_aniso_height_correlated_cos_eval_DG(s, uf, currInteraction, a, a2, ay, ay2);
+				pdf = irr_glsl_beckmann_pdf(currInteraction, uf, a, a2, ay, ay2);
```

Comment:
we can do better than this, see here:
https://github.com/buildaworldnet/IrrlichtBAW/pull/424#discussion_r509445110

This time you're not computing the remainder, so you can compute G1 and G2 expliclty.

You only need to branch for the NDF*G2 and G2/G1 computation, see the implementation functions `wo_clamps` of the ggx, beckmann and phong functions.

the `is_bsdf` check you can move to before and after the partial eval and PDF computation, the BSDF for dielectrics don't differ from the BRDFs except for:
1) they use absolute dot products for NDF and Geometry Functions
2) they use different fresnel (conductor vs side-sensitive dielectric)
3) the fresnel is not part of the PDF
4) the differential factor is slightly different (I provide a function for adjusting it given a BRDF factor)